go/interpreter/lfvm/memory.go

```diff
@@ -68,68 +69,75 @@ func (m *Memory) ExpansionCosts(size uint64) tosca.Gas {
 	return fee
 }
 
-func (m *Memory) EnsureCapacity(offset, size uint64, c *context) error {
-	if size <= 0 {
-		return nil
-	}
+// expandMemoryAndCharge tries to expand memory to the given size.
+// If the memory is already large enough or size is 0, it does nothing.
+// if not enough gas in the context, it returns an error.
+func (m *Memory) expandMemoryAndCharge(offset, size uint64, c *context) error {
+
 	needed := offset + size
 	// check overflow
 	if needed < offset {
 		c.signalError()
 		return errGasUintOverflow
 	}
-	if m.Len() < needed {
-		needed = toValidMemorySize(needed)
-		fee := m.ExpansionCosts(needed)
+	err := m.expandMemoryWithGasFun(needed, func(fee tosca.Gas) error {
 		if !c.useGas(fee) {
 			c.status = statusOutOfGas
 			return errOutOfGas
 		}
-		m.total_memory_cost += fee
-		m.store = append(m.store, make([]byte, needed-m.Len())...)
+		return nil
+	})
+	if err != nil {
+		return err
 	}
+
 	return nil
 }
 
-func (m *Memory) EnsureCapacityWithoutGas(size uint64) {
+// expandMemory expands the memory to the given size without checking gas.
+func (m *Memory) expandMemory(size uint64) {
```

Comment:
There was a reason why the old functions have been called `EnsureCapacity` and `EnsureCapacityWithoutGas`. The idea was that the default should be the safe one, the one doing charging gas and expanding memory as required, while any user would have to explicitly write "WithoutGas" to mark that this operation is side-stepping the gas checks.

With the new naming schema, the critical one is named `expandMemory`, which looks correct in any use case unless the reader recalls that there is also a `expandMemoryAndCharge` that is most likely the one you should use in practically every use case that is not test infrastructure.

For those reasons, I think it should be explicitly warned that `expandMemory` is probably not what you want. Maybe by calling it `expandMemoryUnsafe`.

Reply:
mmmm yes, I see your point. What about renaming old `EnsureCapacity` to `expandMemory` and `EnsureCapacityWithoutGas` to `expandMemoryWithoutCharging` ? I rather avoid the `WithoutGas` part, since it needs to add the fee into `m.total_memory_cost`